docs/specifications/tests/Zone-TP/zone09.md

```diff
@@ -1,47 +1,278 @@
+       1. If the response has the TC flag set, re-query over TCP and use that
+          response instead.
+       2. If there is no DNS response, then add the name server IP to the
+          *No Response MX Query* set.
+       3. Else, if the RCODE of response is not "NoError" ([IANA RCODE List]),
+          then add the name server IP and the RCODE to the
+          *Unexpected RCODE MX Query* set.
+       4. Else, if the AA flag is not set in the response, then add the name
+          server IP to the *Non-authoritative MX* set.
+       5. Else, if there is no MX record with matching owner name in the answer
+          section, then add the name server (IP) to the *No MX RRset* set.
+       6. Else do:
+          1. Extract the MX RRset from the response.
+          2. Add the name server IP and the MX RRset to the *MX RRset* set.
+
+6.  If the set *No Response MX Query* is non-empty, then output
+    *[Z09_NO_RESPONSE_MX_QUERY]* with the name server IP addresses from the set.
+
+7.  If the set *Unexpected RCODE MX Query* is non-empty, then for each RCODE
+    in the set, do:
+    * Output *[Z09_UNEXPECTED_RCODE_MX_RESPONSE]* with the RCODE value
+      ([IANA RCODE List]) and the name server IP addresses from the set.
+
+8.  If the set *Non-authoritative MX* is non-empty, then output
+    *[Z09_NON_AUTHORITATIVE_MX_RESPONSE]* with the name server IP addresses from
+    the set.
+
+9.  If both *No MX RRset* set and *MX RRset* set are non-empty then:
+    1. Output *[Z09_INCONSISTENT_MX_RRSETS]*.
+    2. Output *[Z09_NO_MX_FOUND]* with the name server IP addresses from the
+       *No MX RRset* set.
+    3. Output *[Z09_MX_FOUND]* with the name server IP addresses from the
+       *MX RRset* set.
+
+10. If the *MX RRset* set is non-empty, then do:
+    1. If the RRsets in *MX RRset* are not equal for all name servers then do:
+       1. Output *[Z09_INCONSISTENT_MX_RRSETS]*.
+       2. For each RRset in *MX RRset*, output *[Z09_MX_DATA]* with the mail
+          target from the RDATA and the associated name server IP addresses in
+          the set.
+    2. Else do:
+       1. If the mailtarget of any of the MX records in the RRset in *MX RRset*
+          is "." ("[Null MX][RFC 7505#section-3]") then do:
+          1. If there are more than one record in the MX RRset, then output
+             *[Z09_NULL_MX_WITH_OTHER_MX]* with the mail targets from the RDATA
+             of MX records.
+          2. If the preference of the "Null MX" is non-zero then output
+             *[Z09_NULL_MX_WITH_NON_ZERO_PREFERENCE]*.
+       2. Output *[Z09_TLD_EMAIL_DOMAIN_HARMFUL]* if *Child Zone* is a TLD
+          (top-level domain), unless the MX RRset is a single
+          "[Null MX][RFC 7505#section-3]" record (mail target is ".").
+       3. Output *[Z09_ROOT_EMAIL_DOMAIN_MEANINGLESS]* if *Child Zone* is the root
+          zone, unless the MX RRset is a single "[Null MX][RFC 7505#section-3]"
+          record (mail target is ".").
```

Comment:
These two steps are somewhat complex.
* Both could be shortened by factoring out the "unless the MX RRset is a single Null MX record" condition.
* It would be nice if they used the prefix-form of `if` like all the other outputting steps.

Reply:
Updated wording.